### ISSUE TITLE
Add a combined 2D/3D viewport tab next to (2D 3D Script AssetLib) to allow editing and previewing 2D and 3D nodes together

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#6935]
### Describe the project you are working on

This applies to any 3D project with 2D elements like UI, HUDs, or post process effects (so pretty much all 3D projects)

### Describe the problem or limitation you are having in your project

There are hardly any 3D projects _without_ 2D elements. By 2D I mean those rendered in the 2D viewport, so any node extending CanvasItem. 

The problem is: 
**In order to preview your 3D and 2D elements together, you always have to run the project.** This gets a lot worse whenever you have to edit your 2D and 3D elements together in correspondence with each other. For example shift the position of an object or change it's color - run the scene to see how it looks together, shift and and change color again because it was not right - run the scene again ... repeat a hundred times.

This is terribly annoying for small projects and an impediment and pointless time sink at best for any serious work. [(More examples outlined here)](https://www.reddit.com/r/godot/comments/13pm6ce/comment/jla8708/?utm_source=share&utm_medium=web2x&context=3)

If you try to work around this by creating a 2D utility scene with SubViewportContainer and SubViewports to render your 3D scene in 2D, you have similar problems: It this does not provide nearly the same usability as the 2D/3D viewports combined since the 3D scene has to be rendered by the scenes Camera3D node rather than the editor 3D viewport camera. You also still can't select edit of move any 3D object while also seeing the 2D elements since the editor immediately jumps to the 3D viewport when editing any 3D node. Selecting any 3D object through the 2D viewport is out of question unless the Godot user is determent to dive into weeks or months of custom tooling via EditorPlugin.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

New editor viewport tab:
![image](https://github.com/godotengine/godot-proposals/assets/47016402/98558df7-7523-47a0-8c7d-0639adbfca7a)

Allows previewing, inspecting and editing both 2D objects in the 2D+3D editor viewport:
![image](https://github.com/godotengine/godot-proposals/assets/47016402/1e879045-e391-4177-8588-d6bde5c8ebd0)

as well as previewing, inspecting and editing 3D objects in the 2D+3D editor viewport:
![image](https://github.com/godotengine/godot-proposals/assets/47016402/4adc03cc-dc36-4b3d-9f64-3567043d6914)

The rendered result for the 2D+3D viewport should be the same whether you have 2D nodes in a scene with a 3D root node or 3D nodes in a scene with a 2D root node. (= as it currently is when running a scene)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

see mock-ups above

### If this enhancement will not be used often, can it be worked around with a few lines of script?

no

### Is there a reason why this should be core and not an add-on in the asset library?

It is a much needed improvement to usability other engines provide out of the box.

**Comment by Zireael07** [comment]
I think it's been bandied about in editor UI/UX improvements proposals, but doesn't have separate issue by itself.

**Comment by AThousandShips** [comment]
Related to #6713

**Comment by golddotasksquestions** [comment]
@Zireael07 
I did quite a bit of issue plowing before starting to wirte this, but I could not find a similar proposal. Would you mind linking to the UI/UX improvement proposal you mean in your comment for easier navigation of the issues?

**Comment by AThousandShips** [comment]
Duplicate of #2635?

**Comment by golddotasksquestions** [comment]
> Duplicate of https://github.com/godotengine/godot-proposals/issues/2635?

@AThousandShips 

Similar problem, but just previewing as is proposed there is not enough. Users need to be able to not just preview both 2D and 3D but also edit both types of nodes at the same time.

**Comment by AThousandShips** [comment]
I think blending both editing together might create more issues than help tbh (also the nature of the root of the scene isn't relevant for this)

**Comment by golddotasksquestions** [comment]
@AThousandShips What issues exactly? If you want to preview select or edit only 2D or only 3D objects you can still do so in their respective viewport tabs.

**Comment by AThousandShips** [comment]
Well the 3D and 2D editors have fundamentally different navigation needs, also the usefulness of having 2D overlays when you're doing 3D editing as opposed to previewing is questionable to me, the annoyance of accidentally selecting something far away in 3D because you missed the 2D element you wanted to click, etc.

**Comment by golddotasksquestions** [comment]
It seems to me you are missing the point.

Navigation is not an issue. We could decide for example to 

a) not have 2D navigation in this 2D+3D editor viewport at all, instead just display all 2D nodes on top with a bit of margin (see mock-ups in the post above). This is essentially the same as having a CanvasLayer with GUI children.
b) have 3D navigation as default but switch to the 2D navigation when pressing a modifer key
c) have 2D navigation as default but switch to 3D navigation when pressing a modifer key

Personally I would like this to be configurable in the Editor Settings.

> the annoyance of accidentally selecting something far away in 3D because you missed the 2D element you wanted to click, etc.

It should work just like it currently works: 2D is always rendered on top of 3D elements in Godot. Therefore if you click a 2D element, the 2D element is selected, if you click somewhere where there is no 2D element, the 3D element is selected if there is one.
However I would not expect many users to use the viewport to select objects in the viewport. Viewport selection already sucks terribly in both 2D and 3D editor viewports and is mostly a gimmick if you ask me. This is not going or meant to solve or improve this.

I don't know about you, but I quickly learned selecting objects in Godot is best done through the scene panel. 

The main benefit is to be able to preview select (to inspect properties, most of the time through the scene panel) and edit both of them at the same time **while seeing what you are doing** and what effect it has in relation to both your 2D as well as 3D elements.

> also the usefulness of having 2D overlays when you're doing 3D editing as opposed to previewing is questionable to me,

If you don't have a use for this feature and don't know a solution for the problem other have either, then why are you commenting? What you are doing is not constructive.

**Comment by AThousandShips** [comment]
> you are wasting everyone's time saying exactly nothing.

Since you are unable to engage fairly and constructively in the face of constructive criticism I'll just leave, please refer to the [code of conduct](https://godotengine.org/code-of-conduct/)

**Comment by KoBeWi** [comment]
Related: https://github.com/godotengine/godot/pull/51817

**Comment by aaronfranke** [comment]
@golddotasksquestions That would be confusing though, how would you know if you were editing 2D or 3D items without having it explicitly selected? How would the toolbars at the top below the scene tabs handle this? How could you discern trying to deselect from 2D by clicking on an empty space from performing a raycast into the 3D scene?

I think the only reasonable solution is what https://github.com/godotengine/godot/pull/51817 is proposing, which was actually accidentally happening back when the Vulkan branch was in early development https://github.com/godotengine/godot/issues/40553

**Comment by golddotasksquestions** [comment]
@aaronfranke 

I really have no idea why you would think this would be confusing. It would behave exactly like the existing 2D and 3D viewports: If you select a node it is highlighted in the Scene Panel and in the center editor viewport and you can inspect it's properties in Inspector panel (see mock-ups above). Of course the toolbar also shows the context of the selected node. So exactly the same as it is right now in the 2D and 3D editor viewports. If none of your nodes is selected also exactly is happening: The nodes are no longer highlighted and the Inspector panel is empty.

Personally I quickly got used to only selecting nodes in the Scene Panel unless the scene is almost empty. Because Godot's mouse selection detection via the 2D or 3D viewport really sucks once you have a few overlapping objects in your scene. Same would be true here in the combined 2D+3D viewport.

Deselecting also would work exactly the same as it is now in the existing 2D and 3D viewports: If your scene is nearly empty, you press "Esc" or  click an empty area of the Scene Panel or click in an empty area in the viewport. If your editor viewport is filled with elements top to bottom so you don't have an empty spot to click, either press the "Esc" key or click an empty area in the Scene Panel. If the Scene Panel is also completely filled with nodes, still pressing "Esc" will deselect your current selection. So once again, exactly the same behavior, problems, usability and solutions we have now with the 2D and 3D viewports. "Esc" always works.

So I really don't understand what would be more confusing about this than what we have now. If anything if is _**less confusing**_ because users would finally not have their nodes suddenly disappear when they switch to selecting a different type of node.

And if they specifically want to only work in either the 3D or 2D context they can still do so in the respective 2D and 3D viewports. Having this combined 2D+3D Viewport to preview edit and select can only be a improvement in usability.

https://github.com/godotengine/godot/pull/51817 is **not** a solution to this issue _at all_. You would still have to change between the 2D and 3D editor viewports if you want to edit either type of node and being in 3D editor viewport won't show you the 2D nodes.

**Comment by Zireael07** [comment]
@aaronfranke That's why I said viewport selecting (i.e. click to select in viewport) would have to be disabled in the combined mode.

That's different from *showing* selections made via other means

**Comment by yosoyfreeman** [comment]
> If you don't have a use for this feature and don't know a solution for the problem others have either, then why are you commenting? What you are doing is not constructive.


He is commenting because is an active Godot contributor who understand the philosophy of the engine and user experience topics pretty well. He can sound a bit too serious sometimes, as me, but is ALWAYS trying to help.

As for this, can't you just run the scene with F6 and edit your 2d scene? Godot supports live editing and is what i do to do this.

> you are wasting everyone's time saying exactly nothing.

Yeah, nope. I disembark here. We have a code of conduct for something. I can not, in good conscience, participate in a conversation where @AThousandShips had to leave because of the constant dehumanization.  I don't think that we, as a community, should continue this conversation like nothing happened.

**Comment by AThousandShips** [comment]
@yosoyfreeman thank you! It's she though thank you

Tbh I think this is a good reson to enable comment reporting on this repository, it's enabled on the engine one

**Comment by yosoyfreeman** [comment]
Sorry. I got angry and forgot to use He/she or they. Is corrected now.

**Comment by AThousandShips** [comment]
Easy mistake, thank you for correcting

**Comment by yosoyfreeman** [comment]
You are totally right. I try to, but i usually forget. Still learning. English is not my native language and here in Spain teach us in school that using "They" for a single person is wrong, so i'm still trying to un-learn that.

PD: I don't think this is offtopic. This issue is happening more as Godot grows and the thing i loved the most about this community is how inclusive it is. It is, or was, a safe space. One of the guidelines is "Assume good intention" Which is not trivial when discussing online. And i say that as an autistic person who find human interaction, specially in this kind of context, pretty hard.

We need more moderators.

**Comment by Zireael07** [comment]
> As for this, can't you just run the scene with F6 and edit your 2d scene? Godot supports live editing and is what i do to do this.

Live editing sounds like a convoluted workaround to many 2d in 3d scenarios (such as placing world space labels, though that specific use case is mostly solved by Label3D in 4.0, there are many other 2D in 3D scenarios much more complex)

**Comment by yosoyfreeman** [comment]
I see. Is a powerful system but not the easiest approach. I'm with you then. Thank you for explaining it to me. 

This is my last comment here until someone with moderation privileges say something. Have a good day!

**Comment by Zireael07** [comment]
>  Is a powerful system but not the easiest approach. I'm with you then

Exactly. I have used 2d in 3d in two different projects and while one case is covered by the Label3D, the other is much more complex (think a diegetic/interactable computer in a 3D world), and it would be MUCH easier if we could edit/preview them in editor immediately, instead of running the game only to discover something got out of viewport/something is clipped etc etc.

**Comment by YuriSizov** [comment]
> We need more moderators.

@yosoyfreeman I think your response was great, and both you and @AThousandShips handled it very well. Thank you for your compassion and efforts to build a better community for Godot.

**Comment by Johnrobmiller** [comment]
![image](https://github.com/godotengine/godot-proposals/assets/55205360/c1f99af5-af83-4360-a85b-46fcef652ce1)

It would be a huge mistake to implement the UI in the screenshot. Good usability principles dictate that an product should present the simplest possible interface that serves the purposes of most number of its users. Niche features should be optionally accessed in a way that is easy to find but doesn't clutter the interface. This principle ensures that the interface doesn't baboon in complexity. However, the proposed implementation violates this principle by adding an extremely niche feature alongside non-niche features core to the program. Would it not be far better to handle this some other way?

**Comment by golddotasksquestions** [comment]
@Johnrobmiller
I agree with you on the principle, but I don't agree with you at all that a combined 2D/3D viewport is a niche feature.
If you do a lot of work between 2D and 3D in the same project, which applies to most 3D games btw, and many 2D games with 3D elements, you will know how essential it is to judge and edit both at the same time. 

Having to always run the scene to see your 2D UI on top of your 3D game is a massive waste of time and terrible user friendliness as you have to move and edit your 2D "blindly" not knowing where your 3D elements are and vice versa.

Even when using SubViewportContainers and Subviewports, making changes which depend on both 2D and 3D is impossible, because whenever you touch 3D nodes the editor viewport switches to "3D" -> without the 2D elements visible.

**Comment by TobiasDn** [comment]
@Johnrobmiller This is far from a niche feature. Having a way to preview the combined results of 2D and 3D elements would be helpful for almost all 3D game development, for previewing the HUD over the 3D scene. In fact, I think this is so commonly relevant that some form of this should be default behavior in 3D, rather than it's own panel.

I don't think the approach suggested here, where the 2D elements are always displayed, is the way to go though. I think this should only happen when previewing a camera, since that is the only scenario where the relation between 3D and 2D elements will meaningfully reflect the situation in the game. Being able to also manipulate 2D elements in that scenario would be great.

**Comment by aaronfranke** [comment]
I would like to point out that the Godot editor can be extended though GDScript, including adding new tabs at the top.

If a combined 2D+3D editor viewport is desired, you can do it without editing the engine. Just make an addon in GDScript.

**Comment by Meta-Ben** [comment]
@aaronfranke speeking of that ! 

I originally made a plugin to allow the preview of the 2D while working on 3D and vice-versa, but with the help of a contributor the plugin evolve and now you can work on 2D AND 3D on a splitted view, ( @golddotasksquestions maybe it can help you )

See the link of the plugin : https://github.com/Meta-Ben/DualEditor
